### PR TITLE
Extended transactions support

### PR DESCRIPTION
**Pull request by LordTermor**
This PR changes how transactions are handled in the code base allowing to start and end transactions at application level. This leads to much faster database access.

### CODE REVIEW

**Comment by LordTermor** [comment]
@romangg should I address some smaller persistence-related issues like https://github.com/anydistro/bxt/issues/63 as a part of this PR? I initially wanted to but this change turned out as a pretty big one.

**Comment by romangg** [comment]
> @romangg should I address some smaller persistence-related issues like #63 as a part of this PR? I initially wanted to but this change turned out as a pretty big one.

Better to put them in a separate pr.

File: daemon/core/domain/entities/Package.cpp [hunk 1]
```diff
@@ -70,12 +70,8 @@ Package::Result<Package> Package::from_file_path(
 
     if (!name.has_value()) { return bxt::make_error<ParseError>(); }
 
-    auto desc = bxt::Utilities::AlpmDb::Desc::parse_package(filepath, false);
-    if (!desc.has_value()) {
-        return bxt::make_error_with_source<ParseError>(std::move(desc.error()));
-    }
-
     Package result(section, *name, false);
+
```
Comment (romangg) [inline daemon/core/domain/entities/Package.cpp:74]: Unneeded new line

File: daemon/core/application/services/AuthService.h [hunk 2]
```diff
@@ -21,15 +22,17 @@ class AuthService {
 public:
     BXT_DECLARE_RESULT(AuthError);
 
-    AuthService(Domain::UserRepository& user_repository)
-        : m_user_repository(user_repository) {}
+    AuthService(Domain::UserRepository& user_repository,
+                Domain::UnitOfWorkBaseFactory& uow_factory)
+        : m_user_repository(user_repository), m_uow_factory(uow_factory) {}
```
Comment (romangg) [inline daemon/core/application/services/AuthService.h:27]: I recommend to use uniform initialization.

File: daemon/core/application/services/PackageLogEntryService.cpp [hunk 3]
```diff
@@ -10,19 +10,37 @@
...
         [this](const Application::Events::SyncFinished &sync_event) {
-            for (const auto &synced_package : sync_event.packages_synced) {
-                m_repository.add(Domain::PackageLogEntry {
-                    synced_package, Domain::LogEntryType::Add});
-            }
+            for (const auto &package : sync_event.packages_synced) {
+                Domain::PackageLogEntry entry(package,
+                                              Domain::LogEntryType::Add);
+                coro::sync_wait(
+                    [](auto *self, const auto entry) -> coro::task<void> {
```
Comment (romangg) [inline daemon/core/application/services/PackageLogEntryService.cpp:26]: Remove asterix. auto is enough. Also could you provide a reference instead? That would be in general better for a not-null argument to a function call.

File: daemon/core/domain/repositories/PackageRepositoryBase.h [hunk 4]
```diff
@@ -23,26 +22,17 @@ struct PackageRepositoryBase : public ReadWriteRepositoryBase<Package> {
...
-        const std::function<bool(const Package& pkg)> predicate) {
-        return coro::sync_wait(find_by_section_async(section, predicate));
-    }
-    virtual TResult find_by_section(const Section section, const Name name) {
-        return coro::sync_wait(find_by_section_async(section, name));
-    }
+    virtual coro::task<TResult>
+        find_by_section_async(const Section section,
+                              const Name name,
+                              std::shared_ptr<UnitOfWorkBase> uow) = 0;
```
Comment (romangg) [inline daemon/core/domain/repositories/PackageRepositoryBase.h:36]: No need to initialize with 0

File: daemon/infrastructure/DeploymentService.h [hunk 5]
```diff
@@ -19,15 +19,16 @@
 #include <vector>
 
 namespace bxt::Infrastructure {
-
```
Comment (romangg) [inline daemon/infrastructure/DeploymentService.h:22]: Unrelated change

**Review by romangg — APPROVED** [review]

File: daemon/core/domain/repositories/PackageRepositoryBase.h (hunk 4 above)
Comment (LordTermor) [reply daemon/core/domain/repositories/PackageRepositoryBase.h:36]: It's a pure virtual function

Comment (romangg) [reply daemon/core/domain/repositories/PackageRepositoryBase.h:36]: The formatting was weird on my phone.
